### PR TITLE
[Bug]: Fix Dependencies not resolved correctly

### LINKED ISSUE #14699: [Bug]: Dependencies don't get resolved correclty when moving Documents
### Expected behavior

- After Moving child-elements from one document tree into another, the dependencies should be correctly updated.

### Actual behavior

The Documents are being moved, yet the dependencies don't change in the parent document of the sourcetree. 
Even though there no longer is  an obvious dependency in the moved elements. (Checking properties and the content itself)  

Might also affect objects and assets.

### Steps to reproduce

In the [demo](https://demo.pimcore.fun/admin/?_dc=1679303801&perspective=) I did the following:

1. Check dependencies "Required by" on the /de Document (ID 136; 34 Documents/Snippets)
2. Move the "Magazin" (Id 139) From the "/de"(ID: 136)  into the "/en" Tree (Id: 3)
3. Check properties in the Magazin (Id: 139) and update the "sidebar" property to reference the snippet in the /en-Tree
4.  Check dependencies on the /de Document again (ID 136; still 34 Documents/Snippets)

The only thing that changed is that all Documents within the "Magazin" are moved to the end of the List.

### PR DESCRIPTION
**Pull request by kingjia90**
## Changes in this pull request  
Resolves https://github.com/pimcore/pimcore/issues/14699

## Additional info
The problem was due the fact that properties are inheritable and are dynamically loaded (not stored in db, getted from parent element) while the dependencies are carbon copied the inherited properties values on element save and not synced again outside the save. 
After PR, the children gets all the dependencies on latest inherited properties.

### <samp>🤖 Generated by Copilot at eea5c27</samp>

This pull request refactors the dependency saving logic for assets, documents, and objects. It extracts the common logic to a new static method `updateDependencies` in the `AbstractElement` class and calls it from the respective `save` methods. This change improves code reusability, consistency, and readability.

### <samp>🤖 Generated by Copilot at eea5c27</samp>

> _`updateDependencies`_
> _saves element relations_
> _autumn leaves fall_

### <samp>🤖 Generated by Copilot at eea5c27</samp>

*  Refactor the dependency saving logic for assets, documents, and objects into a single static method in the `AbstractElement` class ([link](https://github.com/pimcore/pimcore/pull/15280/files?diff=unified&w=0#diff-464ea7ee8142d9960b6349fb040c177d818aefdd437151c3417ff72b04fced26R24), [link](https://github.com/pimcore/pimcore/pull/15280/files?diff=unified&w=0#diff-464ea7ee8142d9960b6349fb040c177d818aefdd437151c3417ff72b04fced26R872-R895))
*  Replace the duplicated code for saving dependencies in the `save` methods of the `Asset`, `Document`, and `AbstractObject` classes with calls to the new static method ([link](https://github.com/pimcore/pimcore/pull/15280/files?diff=unified&w=0#diff-867a334e4cc9a097b215afcb2b1f4949910cb2f2103a2cd997d8699d4d414fc9R588), [link](https://github.com/pimcore/pimcore/pull/15280/files?diff=unified&w=0#diff-867a334e4cc9a097b215afcb2b1f4949910cb2f2103a2cd997d8699d4d414fc9L779-L793), [link](https://github.com/pimcore/pimcore/pull/15280/files?diff=unified&w=0#diff-fa70d4e562bbd6e9db822e92f1733cc1d740c01f0bc36b4e923132b700f5cab6R826), [link](https://github.com/pimcore/pimcore/pull/15280/files?diff=unified&w=0#diff-fa70d4e562bbd6e9db822e92f1733cc1d740c01f0bc36b4e923132b700f5cab6L938-L953), [link](https://github.com/pimcore/pimcore/pull/15280/files?diff=unified&w=0#diff-48235de7b392a4ff9ed273062d750e2732a5b1f8355cd76109abf4c5538a6c45R458-R459), [link](https://github.com/pimcore/pimcore/pull/15280/files?diff=unified&w=0#diff-48235de7b392a4ff9ed273062d750e2732a5b1f8355cd76109abf4c5538a6c45L572-L586))

### CODE REVIEW

File: models/Element/AbstractElement.php [hunk 1]
```diff
@@ -868,4 +869,28 @@ protected function renewInheritedProperties()
         $inheritedProperties = $this->getDao()->getProperties(true);
         $this->setProperties(array_merge($inheritedProperties, $myProperties));
     }
+
+    protected static function updateDependendencies($element)
```
Comment (jdreesen) [inline models/Element/AbstractElement.php]:
```suggestion
    protected static function updateDependendencies(object $element): void
```

File: models/Element/AbstractElement.php [hunk 2]
```diff
@@ -868,4 +869,28 @@ protected function renewInheritedProperties()
...
+    {
+        $type = match (true) {
+            $element instanceof \Pimcore\Model\Asset => "asset",
+            $element instanceof \Pimcore\Model\Document => "document",
+            $element instanceof \Pimcore\Model\DataObject\AbstractObject => "object",
+            default => "object"
```
Comment (jdreesen) [inline models/Element/AbstractElement.php]:
```suggestion
            default => throw new \InvalidArgumentException('Unexpected element');
```

File: models/Element/AbstractElement.php [hunk 3]
```diff
@@ -868,4 +869,28 @@ protected function renewInheritedProperties()
...
+        $d->setSourceType($type);
+        $d->setSourceId($element->getId());
+
+        foreach ($element->resolveDependencies() as $requirement) {
+            if ($requirement['id'] == $element->getId() && $requirement['type'] == $type) {
+                // dont't add a reference to yourself
+                continue;
+            } else {
+                $d->addRequirement($requirement['id'], $requirement['type']);
+            }
```
Comment (jdreesen) [inline models/Element/AbstractElement.php]:
```suggestion
            }

            $d->addRequirement($requirement['id'], $requirement['type']);
```

**Review by jdreesen — APPROVED** [review]

File: models/Element/AbstractElement.php [hunk 4]
```diff
@@ -868,4 +869,28 @@ protected function renewInheritedProperties()
...
         $this->setProperties(array_merge($inheritedProperties, $myProperties));
     }
+
+    protected static function updateDependendencies(object $element): void
+    {
+        $type = match (true) {
+            $element instanceof \Pimcore\Model\Asset => "asset",
+            $element instanceof \Pimcore\Model\Document => "document",
+            $element instanceof \Pimcore\Model\DataObject\AbstractObject => "object",
+            default => throw new \InvalidArgumentException('Unexpected element')
```
Comment (NiklasBr) [inline models/Element/AbstractElement.php]:
```suggestion
            $element instanceof Model\Asset => "asset",
            $element instanceof Model\Document => "document",
            $element instanceof Model\DataObject\AbstractObject => "object",
            default => throw new \InvalidArgumentException('Unexpected element, expected Asset, Document, or AbstractObject')
```

File: models/Document.php [hunk 5]
```diff
@@ -455,6 +455,21 @@ static function (array $doc) use ($oldPath, $newPath): array {
...
             $this->clearDependentCache($additionalTags);
 
+            // if the path changed, refresh the inherited properties before updating dependencies
+            if ($differentOldPath){
+                $this->renewInheritedProperties();
+            }
+            self::updateDependendencies($this);
+
+            // refresh the inherited properties and update dependencies of each children
+            if ($differentOldPath && isset($updatedChildren) && is_array($updatedChildren)) {
```
Comment (kingjia90) [inline models/Document.php:465]: this only works when moving, while it should be also triggered on normal update/save 🤔 
Maybe need a new messenger queue for recalculating dependencies, but worried about potential perfomance drawback, infinite recursion and race conditions. 

See also `additional info` section on PR top post

File: models/DataObject/AbstractObject.php [hunk 6]
```diff
@@ -824,6 +824,21 @@ public function save()
...
 
+            // if the path changed, refresh the inherited properties before updating dependencies
+            if ($differentOldPath){
+                $this->renewInheritedProperties();
+            }
+            self::updateDependendencies($this);
+
+            // refresh the inherited properties and update dependencies of each children
+            if ($differentOldPath && isset($updatedChildren) && is_array($updatedChildren)) {
+                foreach ($updatedChildren as $updatedDocument) {
```
Comment (aryaantony92) [inline models/DataObject/AbstractObject.php]: Change the name to $updatedDataObject

File: models/Asset.php [hunk 7]
```diff
@@ -586,6 +586,21 @@ public function save()
...
 
+            // if the path changed, refresh the inherited properties before updating dependencies
+            if ($differentOldPath){
+                $this->renewInheritedProperties();
+            }
+            self::updateDependendencies($this);
+
+            // refresh the inherited properties and update dependencies of each children
+            if ($differentOldPath && isset($updatedChildren) && is_array($updatedChildren)) {
+                foreach ($updatedChildren as $updatedDocument) {
```
Comment (aryaantony92) [inline models/Asset.php]: Change the name to $updatedAsset

File: models/DataObject/AbstractObject.php (hunk 6 above)
Comment (kingjia90) [reply models/DataObject/AbstractObject.php]: nice catch, however used shorter `object` suffix

**Pull request merged**